pkg/cloud/services/compute/instance.go

```diff
@@ -566,21 +565,25 @@ func (s *Service) AssociateFloatingIP(instanceID, floatingIP string) error {
 }
 
 func (s *Service) InstanceDelete(machine *clusterv1.Machine, openStackMachine *infrav1.OpenStackMachine) error {
-	if machine.Spec.ProviderID == nil {
+	if openStackMachine == nil {
 		// nothing to do
 		return nil
 	}
-
-	parsed, err := noderefutil.NewProviderID(*machine.Spec.ProviderID)
+	instance, err := s.InstanceExists(openStackMachine.Name)
 	if err != nil {
 		return err
 	}
-	if err = deleteInstance(s, parsed.ID()); err != nil {
-		record.Warnf(openStackMachine, "FailedDeleteServer", "Failed to deleted server %s: %v", parsed.ID(), err)
+	if instance == nil {
```

Comment:
I'm not sure if that's safe. Do we know the deleteInstance has 100% completed when the server is gone in OpenStack?

Reply:
#845 would fix :)

Reply:
Ah yes, and I also though incorrectly. deleteInstance *only* deletes the server and either it works or it doesn't. So the `instance == null` check verifies completely that the deleteInstance did what it should do. (It would be different if we would have to do something else after the server deletion in deleteInstance)

so tl;dr, fine for me as is!